synapse/push/httppusher.py

```diff
@@ -106,9 +115,84 @@ def __init__(self, hs: "HomeServer", pusher_config: PusherConfig):
 
         self.url = url
         self.http_client = hs.get_proxied_blacklisted_http_client()
-        self.data_minus_url = {}
-        self.data_minus_url.update(self.data)
-        del self.data_minus_url["url"]
+        self.sanitized_data = {}
+        self.sanitized_data.update(self.data)
+        del self.sanitized_data["url"]
+
+        if "algorithm" not in self.data:
+            self.algorithm = "com.famedly.plain"
+        elif self.data["algorithm"] not in (
+            "com.famedly.plain",
+            "com.famedly.curve25519-aes-sha2",
+        ):
+            raise PusherConfigException(
+                "'algorithm' must be one of 'com.famedly.plain' or 'com.famedly.curve25519-aes-sha2'"
+            )
+        else:
+            self.algorithm = self.data["algorithm"]
+
+        if self.algorithm == "com.famedly.curve25519-aes-sha2":
+            base64_public_key = self.data.get("public_key")
+            if not isinstance(base64_public_key, str):
+                raise PusherConfigException("'public_key' must be a string")
+            try:
+                self.public_key = PublicKey(
+                    unpaddedbase64.decode_base64(base64_public_key)
+                )
+            except Exception as e:
+                logger.warning(
+                    "Failed to unpack public key: %s: %s", type(e).__name__, e
+                )
+                raise PusherConfigException(
+                    "'public_key' must be a valid base64-encoded curve25519 public key"
+                )
+            del self.sanitized_data["public_key"]
+
+    def _process_notification_dict(self, payload: Dict[str, Any]) -> Dict[str, Any]:
+        """Called to process a payload according to the algorithm the pusher is
+        configured with. Namely, if the algorithm is `com.famedly.curve25519-aes-sha2`
+        we will encrypt the payload.
+
+        Args:
+            payload: The payload that should be processed
+        """
+        if self.algorithm == "com.famedly.curve25519-aes-sha2":
+            # we have an encrypted pusher, encrypt the payload
+            cleartext_notif = payload["notification"]
+            devices = cleartext_notif["devices"]
+            del cleartext_notif["devices"]
+            cleartext = json.dumps(cleartext_notif)
+
+            # create an ephemeral curve25519 keypair
+            private_key = PrivateKey()
+            # do ECDH
+            secret_key = private_key.do_exchange(self.public_key)
+            # expand with HKDF
+            zerosalt = bytes([0] * 32)
```

Comment:
```suggestion
            zerosalt = bytes(32)
```
This also gives a 32-long zero byte string. Can't make my mind up if it's preferable from a readability perspective.

Reply:
soru thinks she prefers to explicitly state that the byte array is filled with zeros, if you have a different strong preference, though, she can adapt, of course